From 648434fa929499a7b68cbaa09eaef025630143bc Mon Sep 17 00:00:00 2001
From: Yuan Can <yuancan@huawei.com>
Date: Mon, 18 Jul 2022 13:22:46 +0000
Subject: [PATCH] driver/svm: save svsp_mm in mm and release it in __mmput

ascend inclusion
category: feature
bugzilla: NA
DTS: AR0001TKKL
CVE: NA

-----------------------------------------------------------------

svsp_mm may still be used by hardware(aicore/dvpp) after process
is gone, same as normal mm, thus release should happen when normal
mm is actually released.

Signed-off-by: Yuan Can <yuancan@huawei.com>
---
 include/linux/mm_types.h |  5 +++++
 kernel/fork.c            | 12 ++++++++++++
 2 files changed, 17 insertions(+)

diff --git a/include/linux/mm_types.h b/include/linux/mm_types.h
index e3eaf4587..f1318e251 100644
--- a/include/linux/mm_types.h
+++ b/include/linux/mm_types.h
@@ -618,8 +618,13 @@ struct mm_struct {
 #endif
 	} __randomize_layout;
 
+#if defined(CONFIG_X86_64) && defined(CONFIG_ASCEND_SVSP)
+ #error "the kabi reserve in mm_struct must be changed first"
+#endif
 #if defined(CONFIG_X86_64)
 	KABI_USE(1, struct mm_struct_extend *mm_extend)
+#elif defined(CONFIG_ASCEND_SVSP)
+	KABI_USE(1, struct mm_struct	*svsp_mm)
 #else
 	KABI_RESERVE(1)
 #endif
diff --git a/kernel/fork.c b/kernel/fork.c
index 5ed5b85a9..6ffa99cc3 100644
--- a/kernel/fork.c
+++ b/kernel/fork.c
@@ -1072,6 +1072,10 @@ static struct mm_struct *mm_init(struct mm_struct *mm, struct task_struct *p,
 	mm_init_uprobes_state(mm);
 	hugetlb_count_init(mm);
 
+#ifdef CONFIG_ASCEND_SVSP
+	mm->svsp_mm = NULL;
+#endif
+
 	if (current->mm) {
 		mm->flags = current->mm->flags & MMF_INIT_MASK;
 		mm->def_flags = current->mm->def_flags & VM_INIT_DEF_MASK;
@@ -1135,6 +1139,14 @@ static inline void __mmput(struct mm_struct *mm)
 {
 	VM_BUG_ON(atomic_read(&mm->mm_users));
 
+#ifdef CONFIG_ASCEND_SVSP
+	if (mm->svsp_mm) {
+		exit_mmap(mm->svsp_mm);
+		mmdrop(mm->svsp_mm);
+		mm->svsp_mm = NULL;
+	}
+#endif
+
 	uprobe_clear_state(mm);
 	exit_aio(mm);
 	ksm_exit(mm);
-- 
2.17.1

